File: examples/python/matrix_basics.py
#!/usr/bin/env python3
"""Python counterpart to examples/lua/matrix_basics.lua.

Python's occpy bindings return Eigen matrices / vectors as numpy arrays —
the layout (3xN for Mat3N, column-major in Eigen storage) is the same as
the Lua Mat3N userdata. Use whichever feels natural for the language.

Run:
    python examples/python/matrix_basics.py [molecule.xyz]
(defaults to examples/scf/water.xyz)
"""

import sys
import numpy as np
import occpy


def main():
    # 1. Construction — just plain numpy.
    print("---- construction ----")
    m = np.array([[1, 2, 3], [4, 5, 6], [7, 8, 9]], dtype=float)
    print(m)
    print("shape:", m.shape, "size:", m.size)

    dynamic = np.zeros((2, 4))
    dynamic[0, 2] = 7.5   # 0-indexed in Python, vs Lua's [1][3]
    print("\ndynamic matrix after [0, 2] = 7.5:")
    print(dynamic)

    v = np.array([1.5, 2.5, 3.5, 4.5])
    print("\nvector:", v)
    print("v.sum() =", v.sum(), "  np.linalg.norm(v) =", np.linalg.norm(v))

    # 2. Indexing — 0-based, m[row, col] (numpy) ↔ m[row+1][col+1] (Lua).
    print("\n---- indexing ----")
    print("m[0, 0] =", m[0, 0], " (top-left)")
    print("m[2, 2] =", m[2, 2], " (bottom-right)")

    m[1, 1] = -99
    print("after m[1, 1] = -99:")
    print(m)

    m[0, :] = [100, 200, 300]   # whole-row assignment
    print("after m[0] = [100, 200, 300]:")
    print(m)

    # 3. Use in a real occ call. Mat3N comes back as a (3, N) numpy array.
    print("\n---- in a real binding ----")
    xyz_path = sys.argv[1] if len(sys.argv) > 1 else "examples/scf/water.xyz"
    print("loading:", xyz_path)
    mol = occpy.Molecule.from_xyz_file(xyz_path)
    calc = occpy.XtbCalculator(mol)

    pos = calc.positions     # numpy array, shape (3, N)
    print("positions shape:", pos.shape)
    print("oxygen x =", pos[0, 0])

    # 4. Returned matrices also come back as numpy.
    result = calc.single_point()
    print("\natomic charges:", np.round(result.atomic_charges, 4))
    print("sum (should be ~0):", result.atomic_charges.sum())

    grad = calc.gradient()
    print("\ngradient (3xN):")
    print(grad)

    # 5. No equivalent "to_table" / "from_table" round trip needed in
    # Python — numpy arrays serialize cleanly to JSON via .tolist().
    print("\nas list:", pos.tolist())


if __name__ == "__main__":
    main()

File: examples/lua/matrix_basics.lua
-- examples/lua/matrix_basics.lua
-- A tour of the Eigen matrix / vector userdata exposed to Lua. occ
-- methods that return matrices (positions, gradients, Mulliken charges,
-- elastic tensors, ...) hand back these userdata directly — no copy at
-- the boundary, in-place mutation works, the same `pos[i][j]` syntax
-- you'd write for a nested table.
--
-- Run:
--     occ lua examples/lua/matrix_basics.lua [molecule.xyz]
-- (defaults to examples/scf/water.xyz)

-- 1. Constructing matrices and vectors
print("---- construction ----")
-- LuaBridge3 doesn't auto-overload constructors; nested-table /
-- explicit-size construction goes through `.from_table` / `.from_size`
-- static factories. Bare `occ.Mat3()` still gives a default-constructed
-- instance.
local m = occ.Mat3.from_table({{1, 2, 3},
                               {4, 5, 6},
                               {7, 8, 9}})    -- fixed 3×3
print(m)                                      -- pretty-prints with rows
print("rows:", m:rows(), "cols:", m:cols(), "size:", m:size())

local dynamic = occ.Matrix.from_size(2, 4)    -- empty 2×4 MatrixXd
dynamic:fill(0)
dynamic[1][3] = 7.5
print("\ndynamic matrix after [1][3] = 7.5:")
print(dynamic)

local v = occ.Vector.from_table({1.5, 2.5, 3.5, 4.5})   -- 1-D
print("\nvector:", tostring(v))
print("v:sum() =", v:sum(), "   v:norm() =", v:norm())

-- 2. Index access is 1-based (Lua convention), maps to Eigen's
-- 0-based storage internally. Row-major access: m[row][col].
print("\n---- indexing ----")
print("m[1][1] =", m[1][1], "  (top-left)")
print("m[3][3] =", m[3][3], "  (bottom-right)")

-- In-place mutation through the row proxy writes straight into the
-- Eigen matrix — no copy, no temporary table.
m[2][2] = -99
print("after m[2][2] = -99:")
print(m)

-- Whole-row assignment with a Lua table (same number of cols):
m[1] = {100, 200, 300}
print("after m[1] = {100, 200, 300}:")
print(m)

-- 3. Use one in a real occ call. Mat3N goes by reference — anything
-- you mutate is reflected in the calculator.
print("\n---- in a real binding ----")
local xyz_path = arg[1] or "examples/scf/water.xyz"
print("loading:", xyz_path)
local mol = occ.load_molecule(xyz_path)
local calc = occ.XtbCalculator(mol)

local pos = calc.positions       -- Mat3N userdata: 3 rows × N atoms
print("positions:", pos)
print("number of atoms:", pos:cols())
print("oxygen x =", pos[1][1])

-- 4. Eigen matrices returned from methods also become userdata. The
-- result fields, the gradient, charges — all native types.
local result = calc:single_point()
print("\natomic charges (Vector):", result.atomic_charges)
print("sum (should be ~0):", result.atomic_charges:sum())

local g = calc:gradient()
print("\ngradient (Mat3N):")
print(g)

-- 5. Convert back to a nested Lua table when you really need one
-- (e.g. for JSON serialization). The reverse is `occ.Mat3N.from_table(t)`.
local t = pos:to_table()
print("\nas Lua table:", type(t), "row 1 =", t[1][1], t[1][2], t[1][3])

local pos2 = occ.Mat3N.from_table(t)
print("round-trip pos2[1][1]:", pos2[1][1])
